subprojects/base-services/src/main/java/org/gradle/internal/operations/DefaultBuildOperationQueue.java

```diff
@@ -129,20 +148,52 @@ private static String getFailureMessage(Collection<? extends Throwable> failures
         return "Multiple build operations failed.";
     }
 
+    private class SubmittedWorker {
```

Comment:
I would keep a count of the pending operations, that we increment in `add()` and decrement in the `run()` method of the worker action once complete. Then in `waitForCompletion()` we would block until this counter is 0. We wouldn't wait for the worker actions to complete.

This does several things. Firstly, it's much simpler than trying to coordinate between the waiting thread and a bunch of workers that may or may not be idle and here simpler means more reliable (eg there's a race condition where we check whether a worker is idle and either block waiting for it to complete or cancel it if not, where the worker becomes busy between checking whether it is idle and cancelling it. I'm sure there are others in there).

Secondly, it would help detangle the queue from the workers so that we can more easily move towards a shared set of workers.

Plus it's less garbage.